bazel: use a sysroot for compilation

rockwotj opened the pull request:
This patchset is two fold:

First and foremost we have a hermetic sysroot we compile against for all builds so now system dependencies should not leak into the build, and we can define them from a docker image. This means we can remove some dependencies from the `install-deps.sh` script.

Secondly, wire up a docker image that can run the broker using the [ultra slim distroless series of containers](https://github.com/GoogleContainerTools/distroless?tab=readme-ov-file#why-should-i-use-distroless-images). There is probably more we can do on the packaging front, but this is a good step. I moved the `LD_LIBRARY_PATH` script to a golang binary since bash is not installed on the distroless images. I'm also thinking that we could move completely to the base image that distroless provides (no shared libraries present), then copy in our shared libraries from our sysroot, but I'm punting on that for now.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

rockwotj committed:
bazel: support packaging as docker image

Big picture:

* Use Distroless containers from Google to have a minimal base image
* Since we don't have bash in these containers change our
  LD_LIBRARY_PATH loader script to static Golang binaries
* Mimic the layout/functionality of our existing images

We don't package host shared libraries (glibc, libgcc) however, so we
still need a solution to package those up, ideally with a clang sysroot.

rockwotj committed:
bazel: add a sysroot for a more deterministic build

Provide a sysroot from an ubuntu jammy docker image to provide a
consistent set of headers, and system libraries for us to use to
build Redpanda. This way there should be less differences between
machines and we can remove a couple of required packages for
development.

rockwotj committed:
bazel/rpk: build with purego

rockwotj committed:
bazel: support cross compiling rust

dotnwat commented on MODULE.bazel:
```diff
@@ -270,6 +270,38 @@ crate.annotation(
     gen_build_script = "off",
 )
 
+# ====================================
```
>bazel: support packaging as docker image
>Big picture:

just to clarify, this is the proposal for the docker image packaging that'll eventually be uploaded into docker hub and used by the community/docker-compose/etc...?

rockwotj replied:
It would be a breaking change to use this exact configuration. I think we could offer this specific version as a `redpanda-slim` variant, but I think we'll stick with the existing packaging story for the medium term. Eventually, we should move our existing setup into Bazel, but I don't see that happening for awhile. I actually wanted to do this docker thing so that I could run/update the Wasm tests to not use the nightly build but use the actual docker image at HEAD, and we can orchestrate everything with Bazel.

dotnwat replied:
>so that I could run/update the Wasm tests to not use the nightly build but use the actual docker image at HEAD, and we can orchestrate everything with Bazel.

got it. yeh i wasn't sure how it was going to be used

rockwotj replied:
Yeah I do think we should move to something slimmed down like this eventually. But I should be more clear about the intent of this for sure

dotnwat commented on MODULE.bazel:
```diff
@@ -131,6 +133,16 @@ llvm.toolchain(
         "linux-x86_64": ["http://redpanda-core-toolchain.s3-website-us-east-1.amazonaws.com/llvm-19.1.1-ubuntu-22.04-x86_64.tar.gz"],
     },
 )
+llvm.sysroot(
+    name = "llvm_19_toolchain",
+    label = "@x86_64_sysroot//:sysroot",
+    targets = ["linux-x86_64"],
+)
```
dang this is very cool. does the sysroot provide isolation, or is it merely providing precedence over the default system root (e.g. to pick up other bits installed on the system from bazel/install-deps.sh).

rockwotj replied:
I've not researched this in depth. It's at the end of the day passing `--sysroot` to clang, but anecdotally I've believe it provides isolation because the first pass I did the build failed to find valgrind and xfs headers even tho I have those locally (I was only pulling in glibc and libgcc_s)

dotnwat replied:
ahh, i see--the other deps aren't for the compiler--stuff like bison, ragel, etc..

rockwotj replied:
Well considering my build works locally but fails in CI I think there is still some leaking from the local system happening...

rockwotj replied:
Although the missing header is in the sysroot...

<img width="260" alt="Screenshot 2025-01-03 at 3 34 20 PM" src="https://github.com/user-attachments/assets/08369aa4-e629-4bc0-af19-0070999169c3" />

dotnwat replied:
eek

rockwotj replied:
Ugh it seems clang-tidy doesn't understand this flag or something... Maybe we need to add -isysroot too...

rockwotj replied:
Turns out that clang-tidy in the sandbox doesn't carry the sysroot dependency... I'll have to figure this one out...

rockwotj replied:
https://github.com/erenon/bazel_clang_tidy/pull/71

dotnwat commented on bazel/toolchain/Dockerfile.sysroot:
```diff
@@ -0,0 +1,37 @@
+    libgcc-12-dev \
+    valgrind \
+    xfslibs-dev \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
```
curious why we don't put the rest of the deps into this list?

rockwotj replied:
Which deps? The ones from `bazel/install-deps.sh`? I believe the rest of those are for tools that we need bazel to invoke (ragel, etc) are there others that are compile time?

dotnwat replied:
yep, i realized this when you mentioned above that sysroot was just passed to clang. i was thinking sysroot was a bazel thing, not a compiler toolchain thing.

rockwotj committed:
bazel: default to purego (no CGO)

rockwotj committed:
bazel: upgrade clang-tidy rules

Required to pull in fixes for a custom sysroot

rockwotj committed:
tests/docker: run shfmt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--89da4873-0372-4645-8da4-0352094b4a05-->
<!--TEST RESULTS-->
<!--bb0eb17e-3b83-4b43-8542-8edfc3df7aeb-->
<!--TEST RESULTS-->
<!--4276142c-3c0c-4904-ae0c-97f992b00c09-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60473">build#60473</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60473#01944974-5dff-47fe-8561-0446bc3dc9bd | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_fuzz_admin_ops | ducktape | https://buildkite.com/redpanda/redpanda/builds/60473#019449cf-8f7f-470e-b3c8-864e343eda36 | FLAKY | 5/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60473#019449cc-02e0-49c6-b7d1-cbee8d21fddf | FLAKY | 4/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60520">build#60520</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60520#01944c1a-f2ab-441c-b013-5f42f0b24fb8 | FLAKY | 1/2 |
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60520#01944c1a-f2ac-4fb3-874f-50211e1cdca2 | FLAKY | 1/2 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60520#01944c61-d124-4ab2-81d9-72f02e57b157 | FLAKY | 4/6 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/60520#01944c61-d121-4476-9138-7591b223b6f9 | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60578">build#60578</a></summary>


| test_id                                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit                                                                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/60578#01944edb-0258-4cea-ad25-c854019cfab8 | FLAKY         | 1/2      |
| rptest.tests.enterprise_features_license_test.EnterpriseFeaturesTest.test_enable_features.feature=Feature.oidc.install_license=False.disable_trial=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60578#01944f38-c658-4bce-8759-a47a491d8541 | FLAKY         | 5/6      |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60578#01944f38-c657-48b6-bad1-9e13b001bd78 | FLAKY         | 4/6      |
| rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.topic_delete_installed_snapshots_test                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60578#01944f38-c658-4bce-8759-a47a491d8541 | FAIL          | 0/1      |
</details>
<!--/body-->

rockwotj committed:
bazel: use python toolchain for genrules

Without this one neat trick, genrules are using the system python3 to
run these python scripts (which are chmod +x python files with
 #!/usr/bin/env python3 at the top). This change explicitly uses the
Bazel toolchain. We must switch to the bazel toolchain because after
removing valgrind from the list of system packages in `install-deps.sh`
there is no longer python available in the docker image we build with.
It's better this way anyways.

rockwotj commented:
CI is failing with the reduced set of packages installed in the docker container because there is no longer a system python interpreter. The latest commit fixes our python3 binaries to use the Bazel toolchain instead.

andrewhsu commented on bazel/packaging/bin_wrapper/main.go:
```diff
@@ -0,0 +1,38 @@
+	"syscall"
+)
+
+var (
+	LDLibraryPath = ""
+	BinaryPath    = ""
+)
```
nit: any reason why these vars are exported? if not, move to `main()`

from lint check:
```
golint bazel/packaging/bin_wrapper/
```
getting:
```
bazel/packaging/bin_wrapper/main.go:22:2: exported var LDLibraryPath should have comment or be unexported
```

rockwotj replied:
I can unexport, but they have to be top level vars not inside main so that the `x_defs` can override them.

rockwotj replied:
done

andrewhsu commented on bazel/toolchain/Dockerfile.sysroot:
```diff
@@ -0,0 +1,37 @@
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+# Copy required glibc files
+RUN mkdir -p /output/pkgs/{usr/,}lib/$(uname -m)-linux-gnu /output/pkgs/usr/include /output/pkgs/lib64 \
+    && cp -r /usr/lib/*-linux-gnu/ld-linux-*.so* /output/pkgs/lib64/ \
+    && cp -r /usr/lib/*-linux-gnu/libc.{a,so}* /output/pkgs/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libc_nonshared.a* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libm.{a,so}* /output/pkgs/lib/$(uname -m)-linux-gnu/ \
+    && (if test $(uname -m) = "x86_64"; then cp -r /usr/lib/x86_64-linux-gnu/libmvec.{a,so}* /output/pkgs/lib/x86_64-linux-gnu/; fi) \
+    && cp -r /usr/lib/*-linux-gnu/libdl.{a,so}* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/librt.{a,so}* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libutil.{a,so}* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libpthread.{a,so}* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libgcc_s.so* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/*crt*.o /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/*-linux-gnu/libatomic.so* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/gcc/*-linux-gnu/12/*.o /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/lib/gcc/*-linux-gnu/12/libgcc*.{a,so}* /output/pkgs/usr/lib/$(uname -m)-linux-gnu/ \
+    && cp -r /usr/include/* /output/pkgs/usr/include/
```
this looks like a space optimization step. are the packages installed in line 6 stable enough to depend on copying these specific files over even if they update during a patch? my concern is a patch update in one of the packages changes a file to a symlink.

rockwotj replied:
Yes they are very stable, these things have changed in decades AFAIK and move very slow. It is a space optimization, the full jammy docker container is >100MB and the resulting tarball here is single digit MB.

andrewhsu reviewed (commented):
golang and dockerfiles syntax lgtm, just nits and questions

dotnwat commented:
```
build_tar failed: error executing PackageTar command (from target //bazel/packaging:redpanda_tar) bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar '--output=bazel-out/k8-fastbuild/bin/bazel/packaging/redpanda.tar.gz' '--mode=0555' '--owner=0.0' '--owner_name=.' ... (remaining 5 arguments skipped)
--
  |  
  | Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
  | /usr/bin/env: 'python3': No such file or directory
```

rockwotj commented:
yeah it's a known issue rules_pkg depends on system python.

I am rewriting the packaging step as a single binary rule in go and removing rules_pkg

rockwotj committed:
bazel: remove rules_pkg

Instead just create a small golang binary that creates a tarball
instead. rules_pkg requires a system python binary, and we'd like to not
have that dependency in CI, so instead just do it in golang with a
custom rule.

rockwotj commented:
Force push: rebase to make buildkite happy

rockwotj committed:
bazel/packaging: unexport variables

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60578-->
<!--9368b4ed-748b-43ef-9742-a16bbea9a64d-->
<!--title-->
## Retry command for [Build#60578](https://buildkite.com/redpanda/redpanda/builds/60578)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.topic_delete_installed_snapshots_test
```
<!--/body-->

rockwotj commented:
/ci-repeat 1
release
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.topic_delete_installed_snapshots_test

rockwotj commented:
The ducktape failures are noise/flakes. This change only effects the Bazel build

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -0,0 +1,109 @@
+                # TODO(bazel): figure out how to make this work properly and slim down the
+                # base image even more.
+                # shared_libraries.append(cc_file)
+                pass
```
FWIW I found out this is because we need to `patchelf --set-interpreter` which is what happens in vtools - I'm going to punt on this for now.

dotnwat replied:
sgtm

rockwotj committed:
bazel/packaging: support building a tarball only redpanda + libs

this tarball will conceptually contain only the current cmake build
produces and will be what vtools consumes until Bazel eats it.

rockwotj commented:
Pushed another [commit](https://github.com/redpanda-data/redpanda/pull/24669/commits/9a64cbce4cfba72370dda274bc672975c944fbdd) that will produce the tarball of just redpanda + shared libraries + sysroot libraries (glibc, etc) that the current vtools build can use. TL;DR is `bazel build --config release //bazel/packaging:vtools_input` should get what vtools needs. Please let me know what is missing and we can add binaries/tools/etc to that tarball as needed.

rockwotj commented:
cc: @ivotron and @gousteris

andrewhsu reviewed (commented):
golang and dockerfiles lgtm, leaving this review as comment so automerge does not trigger before @dotnwat gets a chance to look at rest of PR

willing to change to approve review since PR checks are all green if need to merge sooner

dotnwat commented on bazel/packaging/packaging.bzl:
```diff
@@ -0,0 +1,109 @@
+    # Collect all the shared libraries that we built as part of Redpanda.
+    rp_runfiles = ctx.attr.redpanda_binary[DefaultInfo].default_runfiles.files.to_list()
+    for solib in rp_runfiles:
+        # Why the redpanda binary is marked as a runfile of itself? No idea...
```
maybe bazel doesn't allow a target to reference itself as a runfile. so on the off chance you wanted this (e.g. a binary analyzer analyzes itself in a test) then bazel would need to add that in manually?

rockwotj replied:
I can't really follow the logic well enough here to know. Maybe I should ask in Bazel slack...

https://cs.opensource.google/bazel/bazel/+/master:src/main/starlark/builtins_bzl/common/cc/cc_binary.bzl;l=691;drc=09a1617ecc4c8c74d390bba02801976c6229bbb2

dotnwat reviewed (approved):
lgtm. i'm glad we got rid of rules_pkg. the custom stuff is simple to understand and hack on.

rockwotj commented:
> lgtm. i'm glad we got rid of rules_pkg. the custom stuff is simple to understand and hack on.

Agreed!
